nemo/collections/nlp/modules/common/megatron/attention.py

```diff
@@ -1074,141 +1088,132 @@ def forward(
-        # Update attention mask for inference. [b, np, sq, sk]
-        # ==================================================
-
-        if get_key_value:
-            with torch.no_grad():
-                if layer_past is not None:
-                    attention_mask = attention_mask[
-                        ..., attention_scores.size(3) - 1, : attention_scores.size(3)
-                    ].unsqueeze(2)
-                else:
-                    attention_mask = attention_mask[..., : attention_scores.size(3), : attention_scores.size(3)]
-
-        # ===========================
-        # Attention probs and dropout
-        # ===========================
-
-        # attention scores and attention mask [b, np, sq, sk]
-        attention_probs = self.scale_mask_softmax(attention_scores, attention_mask)
-
-        # This is actually dropping out entire tokens to attend to, which might
-        # seem a bit unusual, but is taken from the original Transformer paper.
+            # This is actually dropping out entire tokens to attend to, which might
+            # seem a bit unusual, but is taken from the original Transformer paper.
 
-        if not self.sequence_parallel:
-            with tensor_parallel.random.get_cuda_rng_tracker().fork():
+            if not self.sequence_parallel:
+                with tensor_parallel.random.get_cuda_rng_tracker().fork():
+                    attention_probs = self.attention_dropout(attention_probs)
+            else:
                 attention_probs = self.attention_dropout(attention_probs)
-        else:
-            attention_probs = self.attention_dropout(attention_probs)
-
-        # =========================
-        # Context layer. [sq, b, hp]
-        # =========================
-
-        # value_layer -> context layer.
-        # [sk, b, np, hn] --> [b, np, sq, hn]
-
-        # context layer shape: [b, np, sq, hn]
-        output_size = (batch_size, np, sq, hn)
-
-        # change view [sk, b * np, hn]
-        value_layer = value_layer.view(value_layer.size(0), value_layer.size(1) * value_layer.size(2), -1)
 
-        # change view [b * np, sq, sk]
-        if self.multi_query_attention:
-            attention_probs = attention_probs.view(batch_size, sq * np, -1)
-        else:
-            attention_probs = attention_probs.view(batch_size * np, sq, -1)
-
-        # matmul: [b * np, sq, hn]
-        context_layer = torch.bmm(attention_probs, value_layer.transpose(0, 1))
-
-        # change view [b, np, sq, hn]
-        context_layer = context_layer.view(*output_size)
+            attention_probs = rearrange(attention_probs, 'b np sq sk -> (b np) sq sk')
+            context_layer = torch.bmm(attention_probs, value_layer)
+            context_layer = rearrange(context_layer, '(b np) sq hn -> b np sq hn', np=np)
+            
+        
+        if self.use_long_attention:
+            assert sq == sk, 'Long Attention can only use for self-attention.'
+            nb = (sq // self.local_context)
```

Comment:
need to subtract total_transient tokens or nb may be wrong

Reply:
We already add the transient tokens to `query_layer` and `key_layer` before pass in `CoreAttention`?

Reply:
yes, because we average hidden for creating transient kv, and we don't pass hidden to here